app/Main.hs

```diff
@@ -0,0 +1,204 @@
+import TM2SP
+import SP2GP
+import CFG2TM
+import TM2SymTM
+import TM2SM
+import SM2GR
+import GP2GAP
+import GapFuncWriter
+import ShowInfo
+import Containers (fromList, toList)
+import Data.Maybe (fromMaybe, isJust)
+
+import System.IO
+import Control.Exception (bracket)
+
+type Approach = Bool -> Grammar -> Handle -> IO ()
+
+data Option =
+      InputFile
+    | OutputFile
+    | ErrorFile
+    | FirstApproach
+    | SecondApproach
+    | SecondApproach_a
+    | SecondApproach_b
+    | Quiet
+    | Help
+    deriving (Ord, Eq, Show)
+
+openOutputAnd :: FilePath -> (Handle -> IO a) -> IO a
+openOutputAnd "" action = action stdout
+openOutputAnd fp action = bracket (openFile fp WriteMode) hClose action
+
+firstApproach :: Approach
+firstApproach quiet grammar handle = do
+    let (sm, accessWord, _) = tm2sm $ symTM $ cfg2tm grammar
+        gr@(GR (a, _)) = sm2gr (sm, accessWord)
+        hub = hubRelation accessWord
+        genmap = fromList $ zip (toList a) $ map ((++) "f." . show) [1..]
+    if quiet
+    then
+        hPutStr handle $ showInfo gr
+    else
+        writeGap gr handle genmap hub
+
+secondApproachTemplate ::
+    (TuringMachine -> IO a) ->
+    (a -> IO GroupPresentation) ->
+    Approach
+secondApproachTemplate tm2a a2gp quiet grammar handle = do
+    tm <- boolean2tm grammar
+    a  <- tm2a tm
+    gp <- a2gp a
+    if quiet
+    then
+        hPutStr handle $ showInfo gp
+    else do
+        fmt <- gapFormat gp
+        hPutStr handle fmt
+
+secondApproach :: Approach
+secondApproach = secondApproachTemplate semigroupGamma groupBeta
+
+secondApproach_a :: Approach
+secondApproach_a = secondApproachTemplate semigroupGamma_1 groupBeta_1
+
+secondApproach_b :: Approach
+secondApproach_b = secondApproachTemplate semigroupGamma_2 groupBeta_1
+
+-- |List of options, which might be used for executing algorithm.
+argd :: [ Arg Option ]
+argd = [ Arg { argIndex = InputFile,
+               argName = Just "input",
+               argAbbr = Just 'i',
+               argData = argDataOptional "file_path" ArgtypeString,
+               argDesc = "Full path to file with grammar definition" },
+         Arg { argIndex = OutputFile,
+               argName = Just "output",
+               argAbbr = Just 'o',
+               argData = argDataOptional "file_path" ArgtypeString,
+               argDesc = "Full path to file for printing results" },
+         Arg { argIndex = ErrorFile,
+               argName = Just "error",
+               argAbbr = Just 'e',
+               argData = argDataOptional "file_path" ArgtypeString,
+               argDesc = "Full path to file, where errors should be recorded during parsing" },
+         Arg { argIndex = FirstApproach,
+               argName = Just "first-approach",
+               argAbbr = Just 'f',
+               argData = Nothing,
+               argDesc = "Use first approach" },
+         Arg { argIndex = SecondApproach,
+               argName = Just "second-approach",
```

Comment:
Description of approaches should be more user-friendly.
1. Add a link to the detailed description of an approach.
2. Add a brief description of an approach.
3. Add bride comparison with other approaches. (When should I use this particular approach instead of another one?)